### DISCUSSION TITLE
cloud optimized geotiff (COG) as data source

### DISCUSSION THREAD

**Discussion by wiesehahn** [maplibre/maplibre-gl-js#1246] [General]
Are COGs supported by maplibe gl js? 

I can not find any information which explicitly stated that its either possible or not. 

For Raster source tiles are mentioned and Image source is for unreferenced images I guess.

**Comment by HarelM** [comment]
I don't think so.
Having said that, you can use `addProtocol` for stuff that are not supported out of the box.

**Comment by wiesehahn** [comment]
Is there a reason it is not possible out of the box to add COGs as data source? 

I am not that deep in the topic, but as far as I understand the advantage of COGs is its internal tiling and the capability of serverless streaming among others. So it looks like its made for adding it as source to webmaps.

What am I missing?

**Comment by jimjam-slam** [comment]
I'm still getting my head MapLibre, but if COG support isn't an immediate priority, I'd love to collaborate with others on a plugin to add it using `addProtocol` 🥳

**EDIT:** As I understand it, the [`georaster-layer-for-leaflet` plugin](https://github.com/GeoTIFF/georaster-layer-for-leaflet), alongside the [`georaster` library](https://github.com/GeoTIFF/georaster), can make requests of COGs in a similar fashion (although I haven't taken a good look at them yet).

**Comment by acalcutt** [comment]
There was some information about using COGs in maplibre linked in this interview with MIERUNE ( https://maplibre.org/news/2022-06-10-mierune-announcement/ )

This was the article in Japanese
https://qiita.com/Kanahiro/items/70b3b8b11bd26cbaf30e

Translated to English
https://qiita-com.translate.goog/Kanahiro/items/70b3b8b11bd26cbaf30e?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=en&_x_tr_pto=wapp

**Comment by wiesehahn** [comment]
I would really appreciate to see a simple implementation which displays some COG with maplibre in the examples.

**Reply by frafra** [reply]
It would be great to have a library similar to PMTiles, but for COG; in the meanwhile, here are some resources:
- https://qiita.com/Kanahiro/items/70b3b8b11bd26cbaf30e#typescript
- https://github.com/blacha/lerc-cog-maplibre

**Comment by wiesehahn** [comment]
While there are some examples how to add COGs to maplibre out there I think it would be nice to add an "official" example to the examples (https://maplibre.org/maplibre-gl-js/docs/examples/). 

There are already examples to [add a WMS source](https://maplibre.org/maplibre-gl-js/docs/examples/wms/) and [add a raster tile source](https://maplibre.org/maplibre-gl-js/docs/examples/map-tiles/), but COGs are getting more widely adopted from day to day.

E.g. Open Layers has a whole bunch of examples about COGs (https://openlayers.org/en/latest/examples/index.html?q=cog).

**Reply by HarelM** [reply]
If you build an example, I'll be happy to review it and add it to the examples in the docs site.
Building an example now is super simple, all you have to do is create an html page that displays the example you want to show in the tests/examples folder and create an image for it.
I don't have any COGs data nor the understanding of how to read or parse it, but using addProtocol should be simple I hope. 
We added a few examples and test around addProtocol which should help.
If you are serious about creating an example, I'll be happy to guide you through the process.

**Comment by nicokant** [comment]
Hi, at NINA we are working on COG support on MapLibre, right now it's very simple, doesn't have great performances and assumes that the input COG meets some requirements.
Here is the PR that implements it: https://github.com/NINAnor/nina-map-explorer/pull/4
We are also interested in adding it to the maplibre-gl-js library or alternatively to publish it as an npm package.

**Reply by HarelM** [reply]
Nice! This uses the `addProtocl` if I understand correctly.
I'm looking into simplifying it with promises and abort controller for version 4.
Please let me know how difficult it would be for you to either support both or create a package for version 3 and below and version 4 and above.
Also I'm thinking about introducing the ability to run a method in the worker thread instead of the main thread with addProtocol.
If this is something that might interest you let me know I can see what can be a good API for this.

**Reply by nicokant** [reply]
Yes, this uses `addProtocol`, it's mostly based on PMTiles code.
Is version 4 already available? I'll take a look and let you know!
Would it be possible to include this inside `maplibre-gl-js` or you think it would be better to create an external package?
As I understand from the documentation, part of the processing (especially related to decompression) done by geotiff.js runs inside a worker, so right now it's not a priority :)

**Reply by HarelM** [reply]
There is a 4.0.0-pre.1 release, but it still doesn't have the breaking changes related to `addProtocol`.
Hopefully within the next week or so I'll publish a pre.2 with this change.

**Reply by HarelM** [reply]
- https://github.com/maplibre/maplibre-gl-js/pull/3433

**Reply by wipfli** [reply]
> Would it be possible to include this inside maplibre-gl-js or you think it would be better to create an external package?

I think it would be best if it could be an external package. Maybe this will need some work on MapLibre to expose more functionality to plugins, but we are open to that in general I think...

**Reply by nicokant** [reply]
I've created a separated repository, and I'm trying to implement it as a plugin, it's still a work in progress so it's missing examples, tests and is not published on NPM. I plan soon to add some example and release it on NPM.
https://github.com/NINAnor/maplibre-gl-cog/

As for v4 support, it's definitely something we want to support, but with less priority than publishing on NPM and providing examples/documentation

**Reply by HarelM** [reply]
v4 is breaking `addProtocol` implementation so if you publish it before 4 is released, you'll need to either support both or decide how to deal with the breaking changes. Just saying. :-)

**Reply by nicokant** [reply]
The idea is to support both the protocol versions, I've seen that also maplibre-contour decided to have the same approach.

The package is published on npm: https://www.npmjs.com/package/@ninanor/maplibre-gl-cog

We are open for PRs and feedbacks!

**Comment by nicokant** [comment]
After a few experiments with COGs and Maplibre I would recommend to serve cloud optimized geotiffs using [TiTiler](https://developmentseed.org/titiler/), it's basically a python service that renders COGs on the fly by tile, using this you can render COGs as a zxy tiles. Moreover you don't need to pre-process your raster to apply a specific colormap, titiler supports changing the colormap on the fly as well.

**Reply by wiesehahn** [reply]
can you explain a little more why you do not recommend your previous solution as in maplibre-gl-cog?

**Reply by nicokant** [reply]
The strongest reason is applying colors, with my previous solution you need to create a cloud optimized geotiff pre-applying color to the whole raster, and if you have a single band raster you still need to normalize it and "manually" (I used some gdal scripts) create the other bands to mimic a 4 band PNG. Also the data type of the raster needs to be considered as the result needs to be an integer data type with a specific number of bits. You can also see in the README some examples of how I did that. This requires a lot of computation, and if you need different color maps applied you also will need to pre-compute them and to store multiple files.
The nice thing about TiTiler is that it uses COG ability to fetch single tiles to apply computation on-the-fly that can be cached, so you can have dynamic color maps applied to your rasters, but you can also write custom algorithms that can be very efficient because they work on a single tile rather than the whole raster.
TiTiler also provides a TileJSON endpoint, so using it with maplibre is straightforward and don't require custom protocols or custom code. TiTiler itself doesn't need a database and can be started as a standalone service, we use nginx to actually serve the COG files.

We are using this solution in production, check for example this link: https://maps.nina.no/datasets/NorwegianEcologicalNetworks. In this case the layers are rendered using a custom algorithm that tries to enhance contrast in the colormap to enhance local differences rather than using a global min/max. You can see that zooming in/out change how the colors are applied.

**Reply by reyemtm** [reply]
@nicokant did you look at the openlayers COG implementation? They have very robust COG support. The issue of course with using titiler is having to stand up a whole server (not sure if you have using serverless in your map above or not). OL has support for mosaics, RGB, color ramps which work on each tile, etc. Though it relies heavily on other libraries inside OL for creating the final image.

**Reply by reyemtm** [reply]
It seemed like you were on the right track with your library - at least it would be a start for exploring COGs without a server.

**Reply by nicokant** [reply]
I haven't actually looked at the OL implementation, it could be probably ported to maplibre but it would require probably some effort to make it work.

You can actually fork the repository and work on that if you want, I guess others could also benefit from this :)

Since titiler is actually serveless, and the result of applying colormaps can be cached (also serverside, not only in the browser) and you can easily extend it with custom algorithms using python and all the python libraries to manipulate images and numpy, for our needs it does not worth investing  time to reimplement the same behavior in javascript

**Comment by reyemtm** [comment]
So the titiler instance you are running in the app above is running on serverless? Do you find that the performance meets your needs? I did not check load times for the raster tiles on that site but just curious. It sounds like it is and that performance is not the #1 reason you are using it…

The trick in OL is that it already has functionality for manipulating images and just uses a very similar method to what you have put together here to then feed into that image pipeline.

**Comment by giswqs** [comment]
Leafmap supports visualizing COG and STAC with MapLibre. See this example. 
https://leafmap.org/notebooks/92_maplibre/#cog-layer

**Comment by reyemtm** [comment]
I’ve updated the plugin above to handle COG using the protocol method - are you using a custom protocol or something else in this example?

**Comment by oscarfonts** [comment]
Hi,

FWIW, we just released our first version of "maplibre-cog-protocol" here:

https://www.npmjs.com/package/@geomatico/maplibre-cog-protocol
https://github.com/geomatico/maplibre-cog-protocol

It can be used to display RGB images, DEMs for terrain or hillshading, or single-band rasters colored with any ColorBrewer scheme.

Feedback is wellcome.

**Reply by four43** [reply]
Just a shout out - that works freakin awesome. nice work! Thank you!

**Comment by reyemtm** [comment]
Nice work - is there a way to add transparency to the borders? I can ask that in your repo I guess. In my version I'm just making everything that is 0 transparent to sort of fake it, but it should be possible to use the mask band as well in the COG. Also, I found that when creating the COG, using SPARSE_OK = true causes errors in the rendering, it's an upstream issue with geotiff, just FYI.

**Reply by oscarfonts** [reply]
Yes please let's continue the conversation on the maplibre-cog-protocol repo. BTW mask is something we still don't support.
